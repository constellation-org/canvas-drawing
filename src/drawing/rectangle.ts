type RectangleToDraw = {
  x: number
  y: number
  width: number
  height: number
  backgroundColor?: string
  borderColor?: string
  borderWidth?: number
  opacity?: number
}

const DEFAULT_OPACITY = 1
const DEFAULT_BORDER_WIDTH = 4
const DEFAULT_BORDER_COLOR = '#000'

export const drawRectangle = (
  context: CanvasRenderingContext2D,
  args: RectangleToDraw
) => {
  context.save()

  context.beginPath()
  context.rect(args.x, args.y, args.width, args.height)

  context.globalAlpha = args.opacity ?? DEFAULT_OPACITY

  if (args.backgroundColor) {
    context.fillStyle = args.backgroundColor
    context.fill()
  }

  context.strokeStyle = args.borderColor ?? DEFAULT_BORDER_COLOR
  context.lineWidth = args.borderWidth ?? DEFAULT_BORDER_WIDTH
  context.stroke()

  context.restore()
}
