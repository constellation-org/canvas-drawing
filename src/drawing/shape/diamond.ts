import { ShapeToDraw, drawShape } from './shape'

export const drawDiamond = (
  context: CanvasRenderingContext2D,
  args: ShapeToDraw
) => {
  const path = (context: CanvasRenderingContext2D, args: ShapeToDraw) => {
    context.beginPath()
    context.moveTo(args.x + args.width / 2, args.y)
    context.lineTo(args.x + args.width, args.y + args.height / 2)
    context.lineTo(args.x + args.width / 2, args.y + args.height)
    context.lineTo(args.x, args.y + args.height / 2)
    context.closePath()
  }
  drawShape(context, path, args)
}
