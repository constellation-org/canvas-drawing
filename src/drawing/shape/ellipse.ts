import Shape, { ShapeToDraw } from './shape'

type EllipseToDraw = Omit<ShapeToDraw, 'path'> & {
  x: number,
  y: number,
  width: number,
  height: number
}

export default class Ellipse extends Shape {
  constructor(args: EllipseToDraw) {
    let path = new Path2D()
    path.ellipse(
      (args.x + args.width / 2),
      (args.y + args.height / 2),
      args.width / 2,
      args.height / 2,
      Math.PI,
      0,
      2 * Math.PI
    )
    super({...args, path })
  }
  draw(context: CanvasRenderingContext2D): void {
    super.draw(context) 
  }
}
