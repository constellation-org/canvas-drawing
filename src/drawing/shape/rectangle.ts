import { ShapeToDraw, drawShape } from './shape'

export const drawRectangle = (
  context: CanvasRenderingContext2D,
  args: ShapeToDraw
) => {
  const path = (context: CanvasRenderingContext2D, args: ShapeToDraw) => {
    context.beginPath()
    context.rect(args.x, args.y, args.width, args.height)
  }
  drawShape(context, path, args)
}
