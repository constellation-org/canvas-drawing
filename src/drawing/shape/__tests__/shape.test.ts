import {drawDiamond} from '../diamond'
import {drawEllipse} from '../ellipse'
import { drawRectangle } from '../rectangle'

describe('[Shape]', () => {
  let canvas: HTMLCanvasElement
  let context: CanvasRenderingContext2D
  beforeEach(function () {
    canvas = document.createElement('canvas')
    canvas.setAttribute('width', "100")
    canvas.setAttribute('height', "100")
    context = canvas.getContext('2d') as CanvasRenderingContext2D
  })
  it('should have a red background', () => {
    drawRectangle(context, { x: 20, y: 30, width: 60, height: 40, backgroundColor: 'red'})
    expect(canvas.toDataURL('image/png')).toMatchSnapshot()
  })
  it('should have a solid, dashed and dotted borders', () => {
    drawRectangle(context, { x: 10, y: 5, width: 80, height: 25, borderStyle: 'solid' })
    drawRectangle(context, { x: 10, y: 37, width: 80, height: 25, borderStyle: 'dashed' })
    drawRectangle(context, { x: 10, y: 70, width: 80, height: 25, borderStyle: 'dotted' })
    expect(canvas.toDataURL('image/png')).toMatchSnapshot()
  })
  it('should have a green border', () => {
    drawEllipse(context, { x: 20, y: 30, width: 60, height: 40, borderColor: 'green'})
    expect(canvas.toDataURL('image/png')).toMatchSnapshot()
  })
  it('should have a 10 pts width border', () => {
    drawEllipse(context, { x: 20, y: 30, width: 60, height: 40, borderWidth: 10})
    expect(canvas.toDataURL('image/png')).toMatchSnapshot()
  })
  it('should have a transparent opacity', () => {
    drawRectangle(context, { x: 10, y: 10, width: 80, height: 60, backgroundColor: 'orange'})
    drawDiamond(context, { x: 20, y: 30, width: 60, height: 60, backgroundColor: 'green', opacity: 0.5})
    expect(canvas.toDataURL('image/png')).toMatchSnapshot()
  })
})
