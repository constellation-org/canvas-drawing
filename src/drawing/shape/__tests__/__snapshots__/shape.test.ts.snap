// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Shape] should have a 10 pts width border 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAAAXNSR0IArs4c6QAABc5JREFUeF7tm1notVMUxn/fnTKFEBfKrIyZMhUSmTKTWaQMuZBImaeIKBcUScmYOTKWCGWWjGUMN+bM5VKPb3/5O33f/+z3Pc9+3/X/rFVv5+Ks/ey1n+fs/e699jqLSAvFwKJQ0WQwpCDBfgQpSAoSjIFg4eQMSUGCMRAsnJwhKUgwBoKFkzMkBQnGQLBwcoakIMEYCBZOzpAUJBgDwcLJGZKCBGMgWDg5Q1KQYAwECydnSAoSjIFg4eQMSUGCMRAsnJwhKUgwBoKFkzMkBQnGQLBwcoakIMEYCBZOzpAUpBcDKwBblWd9YB1grYlHwN9PPN8AXwDvAe8Df/XqfcBGUWfIxsCRwHZFhI1MnHxahHkbuB/43IRrg4kkyIbAMcBRwJa2Ec4PpJkjYe4Fvhyoz3m7iSDIocAFwPYjE/IWcDnwxJhxjCnINsBNwK5jErCUvl8EzgI+GCOuMQTRC/l64NgxBtyhTy1j5wLaGAxmQwuyE/A4sOZgI5ytox+Ag4DXZoOpbz2kIGcCN9eHFsrzROCuISIaSpA7gROGGFDDPm4BzmiI/w/0EII8CBzReiAD4d8OnNqyr9aCKPjbDAPQCftD4N2y+9H54Wfgt/LohC7T6X2V8qxeDpVbAFsD2xriEMRJgGZ8E2spiFIdIrCv/Qo8BjwAPNkXZKLdgeXgeXARrQ+sfhzKIHzUp/G0Ni0FUe5Iv86u9l3Zbt7dtWFHf72orwPW7thO7pqhmnV2ayXIfsBTPaK9AbgM+KNH2z5NViqn83N6NN4XeLZHu3mbtBLkBWCPDsEqj6T9vmbVGKbl9VFggw6dPwfs3cG/yrWFIMpJvVnV+2Knp4Gjy8u5QzO7qzYD2hHu0wFZy5aWL5u1EEQJuksqI3we2KvSdyi3l4HdKju7GLiq0rfKrYUg91Tmqb4qL/2h3hdVhAB6r2iLvV5FA21/tQ22WQtBXgd2rIhQh8WHK/zGcFFsWr6m2avALtOcunzfQpCfAB3KptlqwC/TnEb6Xlvhbyv6/tGdKG0hiIJco2IwOlUrmxrRlitBNI2VZp9mh5ST+DS/Mb4/HHioouNX3BdsLWZIbWZXZw7t/yNabZbhDuBk5wBaCKKt4BWVQV4NXFjpO5TbfeVcVNOfagGuqXGs9WkhiO7K36kNoJxZruzg38pVB0NVoCglUmubAp/UOtf4tRBE/erAt2dNAMVHJ/tTxiosKIlCbcFVilRrymN1Ea8Kt5Ug+/dMmeua9CLg66roZ3dSFaRm53E9oJTHUj7Laq0EUZC1L8alDUjXpcr8fmYd7b9gWmrOBk7via8l2XXh9Z8QWgqyWUkyKhXR114Cbi2VhX0x5rbTvf5pM25VlepRAvVjR0CTGC0FUV8HGCsB35hzfasitskrXBVkL7m+1aeyBbog09ZanzuYCNRyrAx1E2stiILWGq33wvJgl3bY0vca7xCCKDCdS3Q+WcimawXdZja1oQTRIPT3AhUsLDRTUYMu0FRw0dyGFESD0Tqu6nIlFheCqbxIZ40uB92ZxjW0IAp2XeBa4PiZIm/fWFUv51Wm4W3RjCHIkuCVYrkR2N02Gg/Q/+7vCJO0aUlQfdRQ/5palmwqVjgfeMajaz+UMWfIZMS6QzmsPF1ySv1GvriV/nOo8p9HAF09j26RBJlLhg5zqtNS9YeEWtXElK6MRbwqS/Q/lbHqwJY5nKiCTAa8eRFmZ2ATYGVgxTnPEsFUD/znnOf3kuLQLaaEUDVJaFsogoQm0RlcCuJk04CVghhIdEKkIE42DVgpiIFEJ0QK4mTTgJWCGEh0QqQgTjYNWCmIgUQnRAriZNOAlYIYSHRCpCBONg1YKYiBRCdECuJk04CVghhIdEKkIE42DVgpiIFEJ0QK4mTTgJWCGEh0QqQgTjYNWCmIgUQnRAriZNOAlYIYSHRCpCBONg1YKYiBRCdECuJk04CVghhIdEKkIE42DVgpiIFEJ0QK4mTTgPU30b+zZaESdtIAAAAASUVORK5CYII="`;

exports[`[Shape] should have a green border 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAAAXNSR0IArs4c6QAABvVJREFUeF7tm3uIVVUUh78zGRVqhqFmf0Rl9qCH5SM0y5KQTNPoYVRUZgaVRFFY4WPmrjvTyyiIKJDCFCvDSispzZCil2Yp0lihZlF/aZO9p5TKObHPuXe898yZmdvM2ufuob3gwoze+9tr/b5Z556zzj4BPpxyIHAqG58MHohjfwQeiAfimAOOpeM7xANxzAHH0vEd4oE45oBj6fgO8UAcc8CxdHyHeCCOOeBYOr5DPBDHHHAsHd8hHohjDjiWju8QD8QxBxxLx3eIB+KYA46l4zvEA3HMAcfS8R3igTjmgGPp+A7xQBxzwLF0fId4II454Fg6vkM8kC44IBwKnFF4HQcMBgYSMpCAgdHPcTQR0kRAU/Qz7CLgGwIaaWErwr4urJ7pR9zskAaG8g/TCBhRgHCCkitfAVsJ2UwvllPL10q6ajLuAGlgCPu5BrgKOF2two6FGglZTsAyhG8zWrPDZaoPJM9lhMwFRlbZkE1AHuGNauZRPSDCmcCTwNgKDPgO2EzAJkK2AD8BzYQ0E9CMRL+D0J+QPgT0gejVn4CzCBkF0eHvmArWeg+4HeHzCt6r/pbsgTzAYP7mUeDaTqpZCTwHvN9qeHfLN8ACxhMyHZjSidwyDmY289jV3WX/y+ezBSKMBlYBA9pJchshiwlYgkRnSfZCojO0GYTcBJzYzkI/AFMRPraXSLlydkCEWcBT7RS2E5iD8EpWhZetI9GJxEPA8anrB9xALupW65ENEGEpcH1KNebYb75In7BeaSULCHcDtcARKW9fiHBbJTLdeY99IMLLwJUpSb4I3IrwW3cKUP/sw/RjH08XTr+T8osQblZfs0TQLpA4+WdSCrgL4XGbhXVbO8+dhKk5TifueCthD4hEo47PElmbL8lLED6xUo22aD3n0cKKxEmIGb+MQPhSezmjZxPIVuC0kqT/LBSyzUYh1jQbOIX9bcxvRBhmY007QPJcTMjqRMJXIJhri54XwnWFa6IDuQdMJMda7WLsABHeBS4oSXY1wmTt5DPVE94GJpSsuQ4p+10lHX0gEs2kPi3Lrobh1EUjj54bwjnAR4kChiE0ahZlA0geqCtJcg3CJM2kq6YlrAMuLFm/FuF+zXxsAHkhMaeaifCsZtJV08pzCyELS9ZfikRzMbWwAWQjcHZJhmMR1qtlXE0hYRxgpsHF2EB8KFMLG0B+jMbeB6IvQrNaxtUUepAB/FU29NyDtDso7VKm+kBy7CHgyNZsDqE/c/i5S9m59iEzvgfzB1eMHgBE2ADRmD2OGkZThzmM9fyoZwwtZYff9UhFN9gqrl2/Q5KT3QxH1xVX3dU3CjcCi0s+bu7bzOiqXNrn9IHkqSWkvmSxFUjqtFezjmy0hNeAS1sXC5hLLrqPohb6QOJ75cmLwJMQdqhlXQ0hszVpf5sa1OvSB2LMEt4Bxpf4Zv0+gnVGwhIou+ZYizBRe11bQMyV+ZuJZKdUe4tNl82LJw3JeiYQX7mrhh0gcZckx+97gXEIZv9TzwlhOPAhcFhJ0luI/109bAI5uTBkNPujimFu15oxvPpflrozRjDPRYS8BBxeot/MQYyklu021rQHJC5oMmHKTsCA2eR4zEZBapp55hGmDA4DJpFjjdo6CSG7QOJDVwMwP6UAM4Sc5egmh0VRJ7eNHFJ2Sq/OxT6QGIq5LjHba5Lh2jagO4BcYhZXzNlsVxJ1Apl3yIFyphWOx2k17SDgXnK8brvgVP34wtVc4KU99rCPgKuzyi2bDim6UM8oWqLvlOIDNkl/zE4Oc7gw9xn2WIUTbyWdXthKak5A0qKJGiZmebczWyDx4etoYAFEGwc6CtMtS+jFFuZjdr93P4RjIZokzIy2I3UczwP3IOzu/sKVK2QPpJhbPGIxm+XOryBdsxfKPP1kTjW3E0Q70ncTsJsWvkfYyQL6spdB1DCIluiRt6MKL/PXbzZTD4Xo0bjO4n/2OELSjnj88EiGT021B8RsVrgP4a3OiNn8/+p1SFsw5h7K5YXXEJtFl2ibrnuVGla6cs/GHSClBOJtqFOBcwkZTUA/JUC/ABsJ+ICAVdRF4x2nwk0gSYvqOTUCEzKm8H3QF+gdvUJ6twIL+ZWAP6D19Xvhe2cDNWykji+ccj8lmZ4BxHUXFfPzQBTN1JDyQDRcVNTwQBTN1JDyQDRcVNTwQBTN1JDyQDRcVNTwQBTN1JDyQDRcVNTwQBTN1JDyQDRcVNTwQBTN1JDyQDRcVNTwQBTN1JDyQDRcVNTwQBTN1JDyQDRcVNTwQBTN1JDyQDRcVNTwQBTN1JDyQDRcVNTwQBTN1JDyQDRcVNTwQBTN1JDyQDRcVNTwQBTN1JDyQDRcVNTwQBTN1JDyQDRcVNTwQBTN1JD6F6nNRHR89AQ4AAAAAElFTkSuQmCC"`;

exports[`[Shape] should have a red background 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAAAXNSR0IArs4c6QAAAhlJREFUeF7tm7FuQgEMA+H/P7pdGJEsi9hy6bFinHD3wsbzwWuKwHNqG5Z5IGTsIUAIQsYIjK3DhSBkjMDYOlwIQsYIjK3DhSBkjMDYOlwIQsYIjK3DhSBkjMDYOlwIQsYIjK3DhSBkjMDYOlwIQsYIjK3DhSBkjMDYOlwIQsYIjK3DhSBkjMDYOlwIQsYIjK3DhfwDIT9j3zG9zulDfVr2+uYI+eARQMgH8F4fPWV4WsaFjNl9J+Tbfr/ePMGnD/VpGUK4kM8JmA1ciAksHUdImrDZjxATWDqOkDRhsx8hJrB0HCFpwmY/Qkxg6ThC0oTNfoSYwNJxhKQJm/0IMYGl4whJEzb7EWICS8cRkiZs9iPEBJaOIyRN2OxHiAksHUdImrDZjxATWDqOkDRhsx8hJrB0HCFpwmY/Qkxg6ThC0oTNfoSYwNJxhKQJm/0IMYGl4whJEzb7EWICS8f/vJA0oIH+0790nJa94Hzbf3SU81OGp2UIUe70+wjRjFTilOFpmdqc9zUBhGhG1QRCqrj1MIRoRtUEQqq49TCEaEbVBEKquPUwhGhG1QRCqrj1MIRoRtUEQqq49TCEaEbVBEKquPUwhGhG1QRCqrj1MIRoRtUEQqq49TCEaEbVBEKquPUwhGhG1QRCqrj1MIRoRtUEQqq49TCEaEbVBEKquPUwhGhG1QRCqrj1MIRoRtUEQqq49TCEaEbVxC/FmEhlUiA8EgAAAABJRU5ErkJggg=="`;

exports[`[Shape] should have a solid, dashed and dotted borders 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAAAXNSR0IArs4c6QAAAxxJREFUeF7tml1y1TAMRtUtAQtgM/wsg2UAq+G9wJZgzGRKm7k3/uTIqpKcvHWsyPI5ljO3yYNxlSLwUKoaijGEFNsECEFIMQLFyvF0yJ9itR+tHIm1FLSsHCH7toDEWgpCyD4Ty90SaykIIccQ4pEZsqKDJVkf8RIvKehOh3juPRjLkHIREoIxLglC4liGZEJICMa4JAiJYxmSCSEhGOOSICSOZUgmhIRgjEuCkDiWIZkQEoIxLglC4liGZEJICMa4JAiJYxmSCSEhGOOSICSOZUgmhIRgjEuSLiSu9Gtkkt4fSUELLz5y2LdxJNZSEEL2mVjullhLQQipKSSkKpJsE/B0CCwTCCAkAbJnCoR4aCXEIiQBsmcKhHhoJcQiJAGyZwqPEOWX+ncz+7RRQBv7KhT4zsx+bcS1sTedPC2m5bl3tbFHoZZWc1vXvauNfRDySKylIMcPQ4TcNyOxloIQQodwZAlnICEJBDxHVkI5TOER8n6F6wf44gl4hAy9kowv+dwZETLP79AGRghC5hEolpkOQUgxAsXKoUMQUoxAsXLokKsJ+bJa8PrvYjxevZzpHfLqKzxYAUMb2PPD8GA8jlkuQop5Q8iBhSgfOfQEKzmu+F7+aVv0AD7fPwrMXj4lB0LErlVgIuQ2zN6nRKU75JuZfd7YJB/NrMX0rrdm9nsj6KeZtZitK+vbriEhPQCMBxDoHTEBU5DCQwAhHloJsQhJgOyZAiEeWgmxCEmA7JkCIR5aCbEISYDsmcIjpPfChfFt8hJrKWiZB+Avga/Z9f61JLGWghDyj4B3Q677RWItBSGkphDPs4nYQQKeDhmcgts8BDxCemeoZ15i7xBASLGtgZB5QoZOFIQcWMi80s+ZeXqHnBPbvFUhZB7bvMyeZ0heVReeySNkqAUvzHZo6QgZwjbvJoTMYzt0oiDkwELmlX7OzNM75JzY5q0KIfPY5mX2PkN675EZf+nueZdIrKWgZY5bLdhrS8b/C5JYS0EIeaK65wSQWEtBCBkS0jsdbj6YPELynmwXngkhxeQjBCHFCBQrhw4pJuQvQWnrZYAqcq4AAAAASUVORK5CYII="`;

exports[`[Shape] should have a transparent opacity 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAAAXNSR0IArs4c6QAAA4JJREFUeF7t3L9uE0EQBvDPQrElQwcFPS1ISRpoeI9IvEAeKS+AlPegIQ2xBC19CujAUmyhoOV8yD7uzNz+/WZv3ERRdtdz87uZu+xZnsFeVBmYUUVjwcBAyE4CAzEQsgyQhWMVYiBkGSALxyrEQMgyQBaOVYhikAey2LWFIzr5RYN2R24gYaeAKNeiQQYSJrGbLcq1aJCBKAB5eB8lyGoXmb3759BEJ79oUF+FGMjxc8lAyGrNQAyELANk4ViFGAhZBsjCsQoxELIMkIVjFWIgZBkgC8cqxEDIMkAWjlWIgZBlgCwcqxAD4cnA9ecmlotXPDFNtkIcxtWqgbg85UGZJMg+RlsbLCiTA+nDYEKZFMgxDBaUyYD0YZzMG4bt5vCiXrJ9TQKkD2O+AE7fNBCrj8DmngOlepBjGMvHDcL6Jw9K1SASjLYuWFCqBRmDwYRSJYgPBgtKdSAhGAwoVYHEwCiNUg1ITIySKFWApMAohaIeJCVGCRTVIDkwcqOoBcmJkRNFJUgJjFwo6kBKYuRAUQXCgJEaRQ0IE0ZKFBUgjBipUOhBmDFSoFCD9D52XQBnr4Hlk8MnfKV/W/8Abm+AbeCTR1oQDZXRPQliPOSiBNGIEat90YFoalNDbTKkfVGBaK6MWO2LBmQQg/ACLr2BcJWyuhn3ESNqEHcn9fKc745qDMiXT4CD2X8d+yAeDYgL2FoWQAVyFEVR6/JpVW0F0YFor5TQ/0UoQbSihGK446YF0da+QtrU/gU/O4j0DuXvOLdn9awz6xGA5wBORq+WZsIWwB2AX53lvwHo3GF5BCD6OkXRoN2bh39vLzNKWow/XUyCKBoUDcQtxIiSHoMYhA0lD0YSEEnFyce8wAUWuDyYMMcWb3GLp1jLFwoY+R1LfMAZNp2r2D2u8BXXASt7Tx3TsrzfZHBiSRRCDHEZxZfYW7EECikGB4iLIicKMQYPSC4UcgwukNQoCjD4QFKhKMHgBImNogiDFyQWijIMbpBQFIUY/CC+KEoxdICMRVGMoQdEiqIcQxfI/1Dc38k2Cn22nMpuLvpEPLTN4tYi2rX1OTR9FdIeZR9KNwMFt9B9MfSCDLWvNhNKMXSDDKEoxtAP0kVRjlEHSIvifhZ67BpyzejO1XeXFfPoCdcyEDIUAzEQsgyQhWMVYiBkGSALxyrEQMgyQBaOVYiBkGWALJzf20ecgzScln0AAAAASUVORK5CYII="`;
